near-plugins/src/upgradable.rs

```diff
@@ -1,24 +1,48 @@
-//! # Upgradable:
+//! # Upgradable
 //!
-//! Upgradable implementation inspired by [NEP123](https://github.com/near/NEPs/pull/123).
+//! Upgradable trait inspired by [NEP123](https://github.com/near/NEPs/pull/123).
 //!
 //! To upgrade the contract, first the code needs to be staged, and then it can be deployed.
 //!
-//! ## Default implementation:
+//! ## Default implementation
+//!
+//! Only owner or self can call [`Upgradable::up_stage_code`] and [`Upgradable::up_deploy_code`].
 //!
-//! Only owner or self can call `stage_code` and `deploy_code`.
 //! There is no timer or staging duration implemented by default.
 //!
-//! ## Security concerns:
+//! ## Permissions
 //!
-//! Only authorized account is allowed to call `stage_code` and `deploy_code`. There may be several
-//! reasons to protect `deploy_code`. One such reason to keep in mind, is when the code is upgraded
-//! in such a way that requires some sort of migration or initialization. In that case, it is
-//! recommended to run a batched transaction where `deploy_code` is called first, and then a
+//! Only an authorized account is allowed to call [`Upgradable::up_stage_code`] and
+//! [`Upgradable::up_deploy_code`]. There may be several reasons to protect `deploy_code`. For
+//! example if an upgrade requires migration or initialization. In that case, it is recommended to
+//! run a batched transaction where [`Upgradable::up_deploy_code`] is called first, and then a
 //! function that executes the migration or initialization.
 //!
-//! After the code is deployed, it should be removed from staging. This will prevent an old code
-//! with a security vulnerability to be deployed, in case it was upgraded using other mechanism.
+//! ## Stale staged code
+//!
+//! After the code is deployed, it should be removed from staging. This will prevent old code with a
+//! security vulnerability to be deployed.
```

Comment:
Is there a method to remove the staged code? If this is something that should happen every time should we make the upgrade implementation remove it automatically?

Reply:
I’m merging now with that paragraph as is, since it’ll be rewritten soon in upcoming PRs (see below).

---

Staged code can be removed by passing an empty vector to `up_stage_code` ([ref](https://github.com/aurora-is-near/near-plugins/blob/c043add4c2a0810872c4326a55c5162bccd3f4ee/near-plugins-derive/src/upgradable.rs#L32-L39)). It’s currently not mentioned in docs and unidiomatic, I think. For that behavior it is expected to be wrapped in an `Option`? In that case I can open an issue to change the signature of `up_stage_code` to take `code: Option<Vec<u8>>`. Also it would be consistent with `up_staged_code` [returning None](https://github.com/aurora-is-near/near-plugins/blob/c043add4c2a0810872c4326a55c5162bccd3f4ee/near-plugins/src/upgradable.rs#L36-L37) if there’s no code staged.

Staged code should be removed automatically after deployment, it was also pointed out [here](https://github.com/AuditoneCodebase/Aurora-audit-review/issues/9). I’ll make a PR for that change soon.

Reply:
Passing in an `Option` to allow manual unstaging sounds good to me. And it's good we'll be adding the automatic unstaging after deploy as well.

Reply:
Now there's #73 to track this.